### DISCUSSION TITLE
Replacing values for frozen models

### DISCUSSION THREAD

**Discussion by adriangb** [pydantic/pydantic#3352] [Feature Request]
This is a feature request to have something like [`inspect.Parameter.replace()`](https://docs.python.org/3/library/inspect.html#inspect.Parameter.replace) to be able to easily make a copy of a frozen Model w/ a replaced value.

This does not change the functionality of `frozen=True` since mutating in place would still be illegal, it is basically syntactic sugar for some ugly `obj = type(obj)(**obj.dict())` and hopefully w/ less pitfalls.

**Comment by JacobHayes** [comment]
I know this is an old discussion, but I think this is possible with `.copy(update={"field": new_value})` (if that's any better).

**Reply by Argmaster** [reply]
Currently, for Pydantic 2.0.0+ one will have to use `.model_copy(update={"field": new_value})`.

**Reply by maximilianreimer** [reply]
Validation does not seem to work in this case, right?

**Reply by JacobHayes** [reply]
Correct - the update flag assumes the data is safe/valid: https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_copy

**Reply by maximilianreimer** [reply]
What is the intended way of making a modified copy of model?

**Comment by kantholtz** [comment]
Hi, just to add another implementation, this is what I use in version 2.7:

```python
from typing import TypeVar
from pydantic import BaseModel

O = TypeVar("O", bound=BaseModel)

def replace(obj: O, **kwargs) -> O:
    """Replace properties of pydantic model instances."""
    return type(obj).model_validate(obj.model_dump() | kwargs)
```

It behaves like [dataclasses.replace](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace) while also validating the requested changes.

**Comment by kietay** [comment]
Bumping this - seems like a big miss to not have type safe "copy with" semantics

**Reply by Viicos** [reply]
Doesn't [`model_copy()`](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_copy) or [`copy.replace()`](https://docs.python.org/3/library/copy.html#copy.replace) work for you?

**Reply by jmar777** [reply]
Per the `model_copy()` documentation:

> the data is not validated before creating the new model. You should trust this data.

I think the point of this request is to have something _like_ `model_copy()`, but with model/field validation for cases where you _don't_ trust the data (e.g., @kantholtz's example above).

**Reply by rmorshea** [reply]
[revalidate_instances](https://pydantic.dev/docs/validation/latest/api/pydantic/config/#pydantic.config.ConfigDict.revalidate_instances) helps, but it comes at the cost of doing extra validation when you don't need it. A `model_validate(..., revalidate_instances=True)` option could allow you to do this on a one-off basis instance of always.

**Comment by KharchenkoDmitriy** [comment]
Hi,
Tried to wrap an custom alternative, because:
- with @kantholtz solution you will lose excluded fields (as they are not serialized)
- base model_copy dont provide validation (which I just basically add on top)

```python
from typing import TypeVar
from pydantic import BaseModel

T = TypeVar('T', bound=BaseModel)

def frozen_model_copy(base_model_ojb: T, update: dict) -> T:
    """
    Update a frozen Pydantic model with a dictionary of updates.
    """
    for key in update:
        attr = getattr(base_model_ojb, key, None)
        if isinstance(attr, BaseModel):
            # If the attribute is a Pydantic model, we recursively update it
            update[key] = frozen_model_copy(attr, update[key])

    return base_model_ojb.model_validate(base_model_ojb.model_copy(
        deep=True,
        update=update
    ))
```

It has similar semantic to `model_copy`, except that its deep (and recursive) by default (you may add it as a param and cover it as well)
